programs/nft-voter/src/instructions/cast_nft_vote.rs

```diff
@@ -0,0 +1,105 @@
+use anchor_lang::prelude::*;
+use anchor_lang::{Accounts};
+use itertools::Itertools;
+use spl_governance_addin_api::voter_weight::VoterWeightAction;
+use spl_governance_tools::account::create_and_serialize_account_signed;
+use crate::{state::*, id};
+use crate::error::NftVoterError;
+
+/// Casts NFT vote. The NFTs used for voting are tracked using NftVoteRecord accounts
+/// This instruction updates VoterWeightRecord which is valid for the current Slot and the target Proposal only
+/// and hance the instruction has to be executed inside the same transaction as spl-gov CastVote
+#[derive(Accounts)]
+#[instruction(proposal: Pubkey)]
+pub struct CastNftVote<'info> {
+    /// The NFT voting registrar
+    pub registrar: Account<'info, Registrar>,
+
+    #[account(
+        mut,
+        constraint = voter_weight_record.realm == registrar.realm 
+        @ NftVoterError::InvalidVoterWeightRecordRealm,
+
+        constraint = voter_weight_record.governing_token_mint == registrar.governing_token_mint
+        @ NftVoterError::InvalidVoterWeightRecordMint,
+
+        constraint = voter_weight_record.governing_token_owner == governing_token_owner.key()
+        @ NftVoterError::InvalidVoterWeightRecordOwner,
+    )]
+    pub voter_weight_record: Account<'info, VoterWeightRecord>,
+
+    /// The token owner who casts the vote
+    #[account(mut)]
+    pub governing_token_owner: Signer<'info>,
+    
+    /// The account which pays for the transaction 
+    #[account(mut)]
+    pub payer: Signer<'info>,
+
+    pub system_program: Program<'info, System>,
+
+}
+
+/// Casts vote with the NFT
+pub fn cast_nft_vote<'a,'b,'c,'info>(ctx: Context<'a,'b,'c,'info,CastNftVote<'info>>, proposal:Pubkey) -> Result<()> {
+
+    let registrar = &ctx.accounts.registrar;
+    let governing_token_owner = &ctx.accounts.governing_token_owner.key();
+
+    let mut voter_weight = 0u64;
+
+    // Ensure all voting nfts in the batch are unique
+    let mut unique_nft_mints = vec![];
+
+    let rent = Rent::get()?;
+
+    for (nft_info, nft_metadata_info, nft_vote_record_info) in ctx.remaining_accounts.iter().tuples() {
+
+        let (nft_vote_weight, nft_mint) = resolve_nft_vote_weight_and_mint(
+            registrar,
+            governing_token_owner,
+            nft_info,
+            nft_metadata_info,
+            &mut unique_nft_mints)?;
+            
+        voter_weight = voter_weight.checked_add(nft_vote_weight as u64).unwrap();
+
+        // Create NFT vote record to ensure the same NFT hasn't been already used for voting
+        require!(
+            nft_vote_record_info.data_is_empty(),
```

Comment:
I don't get how this mechanism prohibits using the same nft to vote twice. It looks like there's no constraint on `nft_vote_record_info`, so couldn't I pass an arbitrary new account every time I call `cast_nft_vote`?

Reply:
Probably the validation of the nft_vote_record_info address is missing -- it must belong to the nft_mint + proposal

Reply:
Oooh, I guess it works because `create_and_serialize_account_signed()` only works if the pubkey matches the seeds!

Reply:
Correct. Added comment to clarify what happens there